diff --git a/scene/gui/popup.cpp b/scene/gui/popup.cpp
--- a/scene/gui/popup.cpp
+++ b/scene/gui/popup.cpp
@@ -7,8 +7,14 @@
 			parent_window = parent_window->get_parent_visible_window();
 			if (parent_window) {
 				visible_parents.push_back(parent_window);
-				parent_window->connect(SceneStringName(focus_entered), callable_mp(this, &Popup::_parent_focused));
-				parent_window->connect(SceneStringName(tree_exited), callable_mp(this, &Popup::_deinitialize_visible_parents));
+				const Callable focus_entered_callable = callable_mp(this, &Popup::_parent_focused);
+				const Callable tree_exited_callable = callable_mp(this, &Popup::_deinitialize_visible_parents);
+				if (!parent_window->is_connected(SceneStringName(focus_entered), focus_entered_callable)) {
+					parent_window->connect(SceneStringName(focus_entered), focus_entered_callable);
+				}
+				if (!parent_window->is_connected(SceneStringName(tree_exited), tree_exited_callable)) {
+					parent_window->connect(SceneStringName(tree_exited), tree_exited_callable);
+				}
 			}
 		}
 	}
@@ -17,8 +23,14 @@
 void Popup::_deinitialize_visible_parents() {
 	if (is_embedded()) {
 		for (Window *parent_window : visible_parents) {
-			parent_window->disconnect(SceneStringName(focus_entered), callable_mp(this, &Popup::_parent_focused));
-			parent_window->disconnect(SceneStringName(tree_exited), callable_mp(this, &Popup::_deinitialize_visible_parents));
+			const Callable focus_entered_callable = callable_mp(this, &Popup::_parent_focused);
+			const Callable tree_exited_callable = callable_mp(this, &Popup::_deinitialize_visible_parents);
+			if (parent_window->is_connected(SceneStringName(focus_entered), focus_entered_callable)) {
+				parent_window->disconnect(SceneStringName(focus_entered), focus_entered_callable);
+			}
+			if (parent_window->is_connected(SceneStringName(tree_exited), tree_exited_callable)) {
+				parent_window->disconnect(SceneStringName(tree_exited), tree_exited_callable);
+			}
 		}
 
 		visible_parents.clear();
